File: backend/my-issue/src/main/java/com/ssafy/myissue/news/dto/NewsCardResponse.java
package com.ssafy.myissue.news.dto;

import com.ssafy.myissue.news.domain.News;

import java.time.LocalDateTime;

// 목록 카드 한 장의 데이터
public record NewsCardResponse(
        long newsId,
        String title,
        String newspaper,
        LocalDateTime createdAt,
        int views,
        String category,
        String thumbnail

) {
    public static NewsCardResponse from(News news) {
        return new NewsCardResponse(
                news.getId(),
                news.getTitle(),
                news.getNewsPaper(),
                news.getCreatedAt(),
                news.getViews(),
                news.getCategory(),
                news.getThumbnail()
        );
    }
    public static NewsCardResponse toCard(HotNewsCandidates n) {
        return new NewsCardResponse(
                n.getId(),
                n.getTitle(),
                n.getNewsPaper(),
                n.getCreatedAt(),
                n.getViews(),
                n.getCategory(),
                n.getThumbnail()
        );
    }
}

File: android/app/src/main/java/com/ioi/myssue/data/dto/response/NewsCardResponse.kt
package com.ioi.myssue.data.dto.response

import com.ioi.myssue.common.util.TimeConverter
import com.ioi.myssue.domain.model.News
import com.ioi.myssue.domain.model.NewsBlock
import kotlinx.serialization.Serializable


@Serializable
data class NewsContentResponse(
    val type: String,
    val content: String
)

@Serializable
data class NewsCardResponse(
    val newsId : Long,
    val title: String,
    val content: List<NewsContentResponse>,
    val category: String,
    val author: String,
    val newspaper: String,
    val createdAt: String,
    val relativeTime: String,
    val views: Int,
    val url: String,
    val img: String?,
)

fun NewsCardResponse.toDomain(time: TimeConverter) = News(
    newsId = newsId,
    title = title,
    content = content.toDomainBlocks(),
    category = category,
    author = author,
    newspaper = newspaper,
    createdAt = time.toDisplay(createdAt),
    relativeTime = time.toRelative(createdAt),
    views = views,
    url = url,
    img = img,
)

private fun List<NewsContentResponse>.toDomainBlocks(): List<NewsBlock> =
    mapNotNull { dto ->
        when (dto.type.lowercase()) {
            "image" -> NewsBlock.Image(dto.content)
            "desc"  -> NewsBlock.Desc(dto.content)
            "text"  -> NewsBlock.Text(dto.content)
            else    -> null
        }
    }